File: tests/panel_legacy_backfill_test.php
<?php
/**
 * Static + predicate checks for PROMPT-57 legacy relationship backfill.
 * Run: php tests/panel_legacy_backfill_test.php
 */
declare(strict_types=1);

$root = dirname(__DIR__);
$indexPath = $root . '/web/index.php';
$helperPath = $root . '/web/includes/relationship_editor.php';
$ruPath = $root . '/web/lang/ru.php';
$enPath = $root . '/web/lang/en.php';
$daemonPath = $root . '/mail-proxy-daemon.py';

function assert_true(bool $cond, string $msg): void
{
    if ($cond) {
        echo "PASS: {$msg}\n";
        return;
    }
    echo "FAIL: {$msg}\n";
    exit(1);
}

function assert_contains(string $haystack, string $needle, string $msg): void
{
    assert_true(str_contains($haystack, $needle), $msg);
}

function assert_not_contains(string $haystack, string $needle, string $msg): void
{
    assert_true(!str_contains($haystack, $needle), $msg);
}

$index = file_get_contents($indexPath);
$helper = file_get_contents($helperPath);
$ru = file_get_contents($ruPath);
$en = file_get_contents($enPath);
$daemon = file_get_contents($daemonPath);

assert_true($index !== false && $helper !== false, 'read index + helper');
assert_true($ru !== false && $en !== false, 'read i18n');
assert_true($daemon !== false, 'read daemon');

// Task 1 — backlog route + helpers
assert_contains($index, "case 'relationship_backfill':", 'relationship_backfill route');
assert_contains($index, 'renderLegacyRelationshipBackfill()', 'backfill render wired');
assert_contains($index, "action=relationship_backfill", 'nav/link to backfill');
assert_contains($helper, 'function fetchLegacyRelationshipBacklog', 'fetchLegacyRelationshipBacklog');
assert_contains($helper, 'function renderLegacyRelationshipBackfill', 'renderLegacyRelationshipBackfill');
assert_contains($helper, 'relationshipIsLegacyOnly($row)', 'backlog reuses relationshipIsLegacyOnly');
assert_contains($helper, 'function relationshipIsLegacyOnly', 'relationshipIsLegacyOnly still defined');
assert_contains($helper, 'function relationshipMissingFields', 'relationshipMissingFields still defined');

// Task 2 — GET-only prefill into existing form (no second save path)
assert_contains($helper, 'function relationshipExternalClientFormValue', 'prefill helper');
assert_contains($index, 'relationshipExternalClientFormValue($row)', 'form uses prefill helper');
assert_contains($helper, "from=backfill", 'Migrate link uses from=backfill');
assert_contains($index, "(\$_GET['from'] ?? '') === 'backfill'", 'form reads from=backfill GET flag');
assert_contains($index, "data-prefill-source=", 'GET prefill marker on form');
assert_contains($index, "name=\"return_to\"", 'optional return_to after migrate save');
assert_contains($index, "function handleRelationshipSave():", 'single save handler retained');
assert_not_contains($index, 'function handleRelationshipMigrate', 'no separate migrate save handler');
assert_not_contains($helper, '$_SESSION[\'relationship_draft\']', 'no session draft state');
assert_not_contains($index, '$_SESSION[\'relationship_draft\']', 'no session draft in index');

// Task 3 — no bulk migrate / no naming heuristics
assert_not_contains($index . $helper, 'migrate_all', 'no migrate_all endpoint');
assert_not_contains($index . $helper, 'Migrate all', 'no Migrate all UI');
assert_not_contains($helper, 'guessLocal', 'no local-address guess helper');
assert_not_contains($helper, 'inferLocal', 'no local-address infer helper');

// i18n
assert_contains($ru, "'nav.backfill'", 'RU nav.backfill');
assert_contains($en, "'nav.backfill'", 'EN nav.backfill');
assert_contains($ru, "'backfill.count'", 'RU backfill.count');
assert_contains($en, "'backfill.count'", 'EN backfill.count');
assert_contains($en, 'still on the legacy model', 'EN progress sentence');

// Daemon unchanged for this feature
assert_not_contains($daemon, 'relationship_lookup', 'daemon still ignores relationship_lookup');
assert_not_contains($daemon, 'relationship_backfill', 'daemon has no backfill wiring');

// Predicate before/after (Task 4) — load helper functions directly
require_once $helperPath;

$legacyRow = [
    'id' => 101,
    'email' => 'legacy.client@partner.com',
    'external_client_email' => null,
    'local_client_email' => null,
    'local_referent_email' => null,
    'external_account_id' => null,
    'local_client_maildir' => null,
    'active' => 1,
];
$otherLegacy = [
    'id' => 102,
    'email' => 'other@partner.com',
    'external_client_email' => '',
    'local_client_email' => '',
    'local_referent_email' => '',
    'external_account_id' => null,
    'local_client_maildir' => '',
    'active' => 1,
];
$completeRow = [
    'id' => 200,
    'email' => 'done@partner.com',
    'external_client_email' => 'done@partner.com',
    'local_client_email' => 'done@local.loc',
    'local_referent_email' => 'ref@local.loc',
    'external_account_id' => 7,
    'local_client_maildir' => '/var/vmail/local.loc/done/',
    'active' => 1,
];

assert_true(relationshipIsLegacyOnly($legacyRow), 'legacy row is legacy-only');
assert_true(relationshipIsLegacyOnly($otherLegacy), 'second legacy row is legacy-only');
assert_true(!relationshipIsLegacyOnly($completeRow), 'complete row is not legacy-only');

$prefill = relationshipExternalClientFormValue($legacyRow);
assert_true($prefill === 'legacy.client@partner.com', 'prefill carries legacy email');

$blankLocals = relationshipExternalClientFormValue([
    'email' => '',
    'external_client_email' => '',
    'local_client_email' => '',
    'local_referent_email' => '',
    'external_account_id' => null,
    'local_client_maildir' => '',
]);
assert_true($blankLocals === '', 'empty row has empty prefill');

// Simulate backlog filter before migration
$before = [$legacyRow, $otherLegacy, $completeRow];
$beforeLegacy = array_values(array_filter($before, 'relationshipIsLegacyOnly'));
assert_true(count($beforeLegacy) === 2, 'before: 2 of 3 still legacy');

// After operator saves #101 via handleRelationshipSave (four-address filled)
$migrated = $legacyRow;
$migrated['external_client_email'] = 'legacy.client@partner.com';
$migrated['local_client_email'] = 'legacy.client@local.loc';
$migrated['local_referent_email'] = 'ref1@local.loc';
$migrated['external_account_id'] = 3;
$migrated['local_client_maildir'] = '/var/vmail/local.loc/legacy.client/';
$migrated['email'] = 'legacy.client@partner.com';

assert_true(!relationshipIsLegacyOnly($migrated), 'after save: migrated row leaves backlog predicate');

$after = [$migrated, $otherLegacy, $completeRow];
$afterLegacy = array_values(array_filter($after, 'relationshipIsLegacyOnly'));
assert_true(count($afterLegacy) === 1, 'after: 1 of 3 still legacy (N-1)');
assert_true((int)$afterLegacy[0]['id'] === 102, 'after: remaining legacy is the unsaved row');

echo "OK: panel legacy backfill checks passed\n";

File: tests/panel_legacy_backfill_test.py
#!/usr/bin/env python3
"""
Mirror of tests/panel_legacy_backfill_test.php for hosts without php on PATH.
Run: python tests/panel_legacy_backfill_test.py
"""
from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]


def assert_true(cond: bool, msg: str) -> None:
    if cond:
        print(f"PASS: {msg}")
        return
    print(f"FAIL: {msg}")
    sys.exit(1)


def assert_contains(haystack: str, needle: str, msg: str) -> None:
    assert_true(needle in haystack, msg)


def assert_not_contains(haystack: str, needle: str, msg: str) -> None:
    assert_true(needle not in haystack, msg)


def normalize_email(email: str) -> str:
    return email.strip().lower()


def relationship_has_four_address_data(row: dict) -> bool:
    for field in (
        "external_client_email",
        "local_client_email",
        "local_referent_email",
        "local_client_maildir",
    ):
        if str(row.get(field) or "").strip() != "":
            return True
    if row.get("external_account_id"):
        return True
    return False


def relationship_is_legacy_only(row: dict) -> bool:
    return (not relationship_has_four_address_data(row)) and str(row.get("email") or "").strip() != ""


def relationship_external_client_form_value(row: dict) -> str:
    ext = str(row.get("external_client_email") or "").strip()
    if ext:
        return normalize_email(ext)
    if relationship_is_legacy_only(row):
        return normalize_email(str(row.get("email") or ""))
    return ""


def main() -> None:
    index = (ROOT / "web" / "index.php").read_text(encoding="utf-8")
    helper = (ROOT / "web" / "includes" / "relationship_editor.php").read_text(encoding="utf-8")
    ru = (ROOT / "web" / "lang" / "ru.php").read_text(encoding="utf-8")
    en = (ROOT / "web" / "lang" / "en.php").read_text(encoding="utf-8")
    daemon = (ROOT / "mail-proxy-daemon.py").read_text(encoding="utf-8")

    assert_contains(index, "case 'relationship_backfill':", "relationship_backfill route")
    assert_contains(index, "renderLegacyRelationshipBackfill()", "backfill render wired")
    assert_contains(index, "action=relationship_backfill", "nav/link to backfill")
    assert_contains(helper, "function fetchLegacyRelationshipBacklog", "fetchLegacyRelationshipBacklog")
    assert_contains(helper, "function renderLegacyRelationshipBackfill", "renderLegacyRelationshipBackfill")
    assert_contains(helper, "relationshipIsLegacyOnly($row)", "backlog reuses relationshipIsLegacyOnly")
    assert_contains(helper, "function relationshipIsLegacyOnly", "relationshipIsLegacyOnly still defined")
    assert_contains(helper, "function relationshipMissingFields", "relationshipMissingFields still defined")

    assert_contains(helper, "function relationshipExternalClientFormValue", "prefill helper")
    assert_contains(index, "relationshipExternalClientFormValue($row)", "form uses prefill helper")
    assert_contains(helper, "from=backfill", "Migrate link uses from=backfill")
    assert_contains(index, "($_GET['from'] ?? '') === 'backfill'", "form reads from=backfill GET flag")
    assert_contains(index, "data-prefill-source=", "GET prefill marker on form")
    assert_contains(index, 'name="return_to"', "optional return_to after migrate save")
    assert_contains(index, "function handleRelationshipSave():", "single save handler retained")
    assert_not_contains(index, "function handleRelationshipMigrate", "no separate migrate save handler")
    assert_not_contains(helper, "$_SESSION['relationship_draft']", "no session draft state")
    assert_not_contains(index, "$_SESSION['relationship_draft']", "no session draft in index")

    combined = index + helper
    assert_not_contains(combined, "migrate_all", "no migrate_all endpoint")
    assert_not_contains(combined, "Migrate all", "no Migrate all UI")
    assert_not_contains(helper, "guessLocal", "no local-address guess helper")
    assert_not_contains(helper, "inferLocal", "no local-address infer helper")

    assert_contains(ru, "'nav.backfill'", "RU nav.backfill")
    assert_contains(en, "'nav.backfill'", "EN nav.backfill")
    assert_contains(ru, "'backfill.count'", "RU backfill.count")
    assert_contains(en, "'backfill.count'", "EN backfill.count")
    assert_contains(en, "still on the legacy model", "EN progress sentence")

    assert_not_contains(daemon, "relationship_lookup", "daemon still ignores relationship_lookup")
    assert_not_contains(daemon, "relationship_backfill", "daemon has no backfill wiring")

    # Confirm PHP helper source still contains the PROMPT-56 predicates (not re-derived elsewhere)
    assert_true(
        re.search(r"function relationshipIsLegacyOnly\(array \$row\): bool", helper) is not None,
        "PHP relationshipIsLegacyOnly signature present",
    )
    assert_true(
        "return !relationshipHasFourAddressData($row)" in helper,
        "legacy-only still uses relationshipHasFourAddressData",
    )

    legacy_row = {
        "id": 101,
        "email": "legacy.client@partner.com",
        "external_client_email": None,
        "local_client_email": None,
        "local_referent_email": None,
        "external_account_id": None,
        "local_client_maildir": None,
        "active": 1,
    }
    other_legacy = {
        "id": 102,
        "email": "other@partner.com",
        "external_client_email": "",
        "local_client_email": "",
        "local_referent_email": "",
        "external_account_id": None,
        "local_client_maildir": "",
        "active": 1,
    }
    complete_row = {
        "id": 200,
        "email": "done@partner.com",
        "external_client_email": "done@partner.com",
        "local_client_email": "done@local.loc",
        "local_referent_email": "ref@local.loc",
        "external_account_id": 7,
        "local_client_maildir": "/var/vmail/local.loc/done/",
        "active": 1,
    }

    assert_true(relationship_is_legacy_only(legacy_row), "legacy row is legacy-only")
    assert_true(relationship_is_legacy_only(other_legacy), "second legacy row is legacy-only")
    assert_true(not relationship_is_legacy_only(complete_row), "complete row is not legacy-only")

    prefill = relationship_external_client_form_value(legacy_row)
    assert_true(prefill == "legacy.client@partner.com", "prefill carries legacy email")

    before = [legacy_row, other_legacy, complete_row]
    before_legacy = [r for r in before if relationship_is_legacy_only(r)]
    assert_true(len(before_legacy) == 2, "before: 2 of 3 still legacy")

    migrated = dict(legacy_row)
    migrated.update(
        {
            "external_client_email": "legacy.client@partner.com",
            "local_client_email": "legacy.client@local.loc",
            "local_referent_email": "ref1@local.loc",
            "external_account_id": 3,
            "local_client_maildir": "/var/vmail/local.loc/legacy.client/",
            "email": "legacy.client@partner.com",
        }
    )
    assert_true(not relationship_is_legacy_only(migrated), "after save: migrated row leaves backlog predicate")

    after = [migrated, other_legacy, complete_row]
    after_legacy = [r for r in after if relationship_is_legacy_only(r)]
    assert_true(len(after_legacy) == 1, "after: 1 of 3 still legacy (N-1)")
    assert_true(int(after_legacy[0]["id"]) == 102, "after: remaining legacy is the unsaved row")

    print("OK: panel legacy backfill checks passed (python mirror)")


if __name__ == "__main__":
    main()
